### PR TITLE
Add support for slash as sysctl separator to Pod securityContext field and to PodSecurityPolicy

### LINKED ISSUE #102373: Regex for validating Sysctl values in deployments might not be correct.
#### What happened:
The Regex that validates sysctl values in objects doesn't allow names that contain forward slashes. Sysctl values on network subinterfaces contain these forward slashes.

#### What you expected to happen:
Running:
kubectl create -f ./deployment.yaml

Creates the deployment.

#### How to reproduce it (as minimally and precisely as possible):

Create a deployment that brings an additional network interface with subinterface into a pod (using something like Multus). 

Interface will appear as eth2.100

sysctl values for this interface will follow the convention "net.ipv4.conf.eno2/100..."

Attempt to set one of these values in the deployment such as:

spec:
  securityContext:
    sysctls:
    - name: net.ipv4.conf.eno2/100.rp_filter
      value: "0"

It will fail validation, even though it seems like that's the valid name. I wasn't able to escape it.

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.5
- Cloud provider or hardware configuration: On-premise
- OS (e.g: `cat /etc/os-release`): Centos 7
- Kernel (e.g. `uname -a`): 5.5.2-1
- Install tools: kubeadm init
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Starting from Kubernetes version 1.23, the kubelet supports the use of either `/` or `.`
as separators for sysctl names. 
For example, you can represent the same sysctl name as `kernel.shm_rmid_forced` using a period as the separator, or as `kernel/shm_rmid_forced` using a slash as a separator.
For more sysctl parameter conversion method details, please refer to the page [sysctl.d(5)](https://man7.org/linux/man-pages/man5/sysctl.d.5.html) from the Linux man-pages project.

In 1.25, use relaxed validation everywhere, in other words, Pod SecurityContext and PodSecurityPolicy can supports slash as sysctl separator. And I added the corresponding e2e test.

Ref https://github.com/kubernetes/kubernetes/pull/102393#discussion_r645018213
#### Which issue(s) this PR fixes:

Fixes #102373

#### Does this PR introduce a user-facing change?

```release-note
Pod SecurityContext and PodSecurityPolicy supports slash as sysctl separator.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by mengjiao-liu** [comment]
area/test
sig/testing

**Comment by liggitt** [comment]
API bits lgtm, will defer test review to node team

**Comment by mengjiao-liu** [comment]
ping @derekwaynecarr @mrunalp

**Comment by mengjiao-liu** [comment]
Request review @sig-node-approvers

**Comment by mengjiao-liu** [comment]
Request review @sig-node-approvers

**Comment by mengjiao-liu** [comment]
Review review @kubernetes/sig-node-pr-reviews

**Comment by mengjiao-liu** [comment]
ping @derekwaynecarr @mrunalp

File: pkg/apis/policy/validation/validation_test.go
Comment (mrunalp) [inline pkg/apis/policy/validation/validation_test.go:813]: Can we include these two patterns as well?
`
		"a.b/c.d",
		"a/b.c/d",
`

Comment (mengjiao-liu) [reply pkg/apis/policy/validation/validation_test.go:813]: Yes, we can include these two patterns as well.
I'll add them to the test cases.

Comment (mengjiao-liu) [reply pkg/apis/policy/validation/validation_test.go:813]: ping @mrunalp

**Review by mrunalp — APPROVED** [review]

**Comment by mengjiao-liu** [comment]
This PR has been submitted for a long time. We are very near to 1.24 code freeze. The API part has been reviewed and now requires review by sig-node reviewers or approvers. @derekwaynecarr @mrunal @random-liu @dawnchen  @yujuhong @klueska @sjenning

**Comment by mengjiao-liu** [comment]
> API bits lgtm, will defer test review to node team

node team has reviewed. Request approve @liggitt

**Comment by mengjiao-liu** [comment]
Request code review for the e2e section @smarterclayton @johnbelamaric  @derekwaynecarr @tallclair @yujuhong @ehashman @SergeyKanzhelev

**Comment by liggitt** [comment]
are there minimum kernel / container runtime / OS versions required for slash-delimited sysctl support? if so, what is the impact of using a slash-delimited sysctl on a node/kernel/os/container runtime that doesn't support slash-delimited sysctls? is the impact limited to an error running the container, or can it cause problems beyond the scope of that container?

**Comment by ehashman** [comment]
/priority important-longterm
/triage accepted

This probably won't make it for 1.24 but we can try to merge early in 1.25.

File: test/e2e/common/node/sysctl.go
Comment (ehashman) [inline test/e2e/common/node/sysctl.go:183]: This should not be added directly as a conformance test. Let's start with a cluster e2e and consider promoting after we get good signal from the test.

**Comment by sftim** [comment]
/sig security

@mengjiao-liu the PR title describes the context, I think. How would you summarize the _change_ that this PR makes?

**Comment by sftim** [comment]
Eg, you could 
`/retitle Add support for slash as sysctl separator to Pod securityContext field and to PodSecurityPolicy`

**Comment by mengjiao-liu** [comment]
> Eg, you could `/retitle Add support for slash as sysctl separator to Pod securityContext field and to PodSecurityPolicy`

Thanks for your suggestion, it has been changed.

File: test/e2e/common/node/sysctl.go
Comment (mengjiao-liu) [reply test/e2e/common/node/sysctl.go:183]: > Let's start with a cluster e2e

I don't understand what does `cluster e2e test` mean?  Where does it need to be modified?
Do you mean change  `framework.ConformanceIt` to `ginkgo.It` or  test kind?

**Comment by mengjiao-liu** [comment]
> are there minimum kernel / container runtime / OS versions required for slash-delimited sysctl support?

The documentation on sysctl delimiter conversion is not found on the Internet, so I need to check the linux and sysctl source code.


linux kernel support sysctl:
sysctl appeared in `1.3.57` and is fully supported ever since. 

```
/proc/sys
              This directory (present since 1.3.57) contains a number of
              files and subdirectories corresponding to kernel
              variables.  These variables can be read and in some cases
              modified using the /proc filesystem, and the (deprecated)
              [sysctl(2)](https://man7.org/linux/man-pages/man2/sysctl.2.html) system call.

              String values may be terminated by either '\0' or '\n'.

              Integer and long values may be written either in decimal
              or in hexadecimal notation (e.g., 0x3FFF).  When writing
              multiple integer or long values, these may be separated by
              any of the following whitespace characters: ' ', '\t', or
              '\n'.  Using other separators leads to the error EINVAL.
```
ref: 
- https://man7.org/linux/man-pages/man5/proc.5.html
- https://www.linuxjournal.com/article/2365
- source code: https://mirrors.edge.kernel.org/pub/linux/kernel/v1.3/linux-1.3.57.tar.gz

The [sysctl ](https://en.wikipedia.org/wiki/Sysctl) also uses the conversion of slashes and dot separators,at least 10 years ago.

I found the source code of [ procps-ng Sysctl](https://gitlab.com/procps-ng/procps/-/blob/v3.3.0/sysctl.c#L286), the earliest version is 3.3.0 I can found, and the conversion of . and / has been implemented at this time, at least 10 years ago（Maybe earlier, but I can't find the source code further forward）.


> if so, what is the impact of using a slash-delimited sysctl on a node/kernel/os/container runtime that doesn't support slash-delimited sysctls?
is the impact limited to an error running the container, or can it cause problems beyond the scope of that container?

I think that the kernel versions generally used now are greater than version 1.35.7, and the bottom layer of the kernel supports sysctl including slashes.

As far as my own understanding is concerned, the linux kernel finally needs to convert dots into slashes to find files under /proc/sys, so it must support slashes.

The runc code https://github.com/opencontainers/runc/commit/f7d1401a6998bf2036608e049e858ce174651087 only adds namespace verification to sysctl, and then uses syscall to call sysctl in 2015.

The runc code just adds namespace verification to sysctl, and then calls sysctl.

runc 1.1.0 allow slashes in sysctl names, to better match `sysctl(8)`'s behaviour. (https://github.com/opencontainers/runc/pull/3254, https://github.com/opencontainers/runc/pull/3257)

In order to be compatible with runc which does not support slashes to separate namespaces and keys, I have converted the slashes into dot separators before the kubelet calls runc.
https://github.com/kubernetes/kubernetes/blob/1ea07d482a5ec15afac5ea837585174579fc0f57/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L798

So I don't think it will have any effect.
However, I am not very familiar with the linux kernel source code, maybe someone who is more familiar with this part of the code can help to confirm it again?  @liggitt

File: test/e2e/common/node/sysctl.go
Comment (oomichi) [reply test/e2e/common/node/sysctl.go:183]: I think @mengjiao-liu understands @ehashman's comment correctly.
A conformance test needs to be stable as `it is stable and runs consistently (e.g., no flakes), and has been running for at least two weeks` from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements
Then we need to check the stability as `ginkgo.It` for two weeks at least, then the test can promote to a conformance test after that.

File: test/conformance/testdata/conformance.yaml
Comment (pacoxu) [inline test/conformance/testdata/conformance.yaml:2204]: Should we update it to v1.25 for final review?

File: test/e2e/common/node/sysctl.go
Comment (mengjiao-liu) [reply test/e2e/common/node/sysctl.go]: Comment addressed. Request review again. Thanks for @ehashman @oomichi

File: test/conformance/testdata/conformance.yaml
Comment (mengjiao-liu) [reply test/conformance/testdata/conformance.yaml]: ref https://github.com/kubernetes/kubernetes/pull/106834#discussion_r875252364 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements

> it is stable and runs consistently (e.g., no flakes), and has been running for at least two weeks

According to this documentation,I have removed this update from the `test/conformance/testdata/conformance.yaml`. Thanks for the catch.

**Comment by pacoxu** [comment]
/lgtm
move to Done in sig-node board.

**Comment by mengjiao-liu** [comment]
Please review again. @liggitt  @lavalamp

File: test/e2e/common/node/sysctl.go
Comment (liggitt) [inline test/e2e/common/node/sysctl.go:183]: the test isn't marked `[LinuxOnly]` currently, right?

**Review by liggitt — COMMENTED** [review]
/approve
for API changes

/hold for one comment on the e2e test

File: test/e2e/common/node/sysctl.go
Comment (mengjiao-liu) [reply test/e2e/common/node/sysctl.go:183]: https://github.com/kubernetes/kubernetes/blob/7e2c02d4c86d73051b6d4f47298d35a7ffb42b6a/test/e2e/common/node/sysctl.go#L35
The top layer of this file has defined `[LinuxOnly]`.

**Pull request merged**
